### ISSUE TITLE
torch.compile() produces completely different outputs for standard neural networks with single random operation

### ISSUE THREAD

**Issue by Blooming-Tree** [pytorch/pytorch#172212]
### 🐛 Describe the bug

When using `torch.compile()` on a standard multi-layer neural network containing CNN, BatchNorm, LSTM, and linear layers with a single random operation at the output, the compiled model produces completely different numerical outputs compared to normal execution, despite fixed random seeds and evaluation mode. Every output value differs significantly, indicating fundamental issues with deterministic computation in compiled mode.
code：
```
import torch
import torch.nn as nn
import torch.nn.functional as F
import random
CUDA_AVAILABLE = torch.cuda.is_available()
TORCH_CUDNN_SDPA_ENABLED = False

class MultiLayerDNN(nn.Module):

    def __init__(self):
        super().__init__()
        self.conv1 = nn.Conv2d(3, 16, kernel_size=3, padding=1)
        self.bn = nn.BatchNorm2d(16)
        self.relu = nn.ReLU()
        self.sigmoid = nn.Sigmoid()
        self.tanh = nn.Tanh()
        self.fc = nn.Linear(16 * 14 * 14, 64)
        self.lstm = nn.LSTM(64, 32, num_layers=2, batch_first=True)
        self.out = nn.Linear(32, 10)

    def forward(self, x):
        x = self.conv1(x)
        x = self.bn(x)
        x = self.relu(x)
        x = F.max_pool2d(x, 2)
        x = torch.flatten(x, 1)
        x = self.fc(x)
        x = self.tanh(x)
        x = x.unsqueeze(1)
        x, _ = self.lstm(x)
        x = x[:, -1, :]
        x = self.sigmoid(x)
        x = self.out(x)
        x = x - torch.randn_like(x)
        return x

def get_default_model():
    return MultiLayerDNN()

def get_sample_inputs():
    return (torch.randn(8, 3, 28, 28),)

def main():

    torch.manual_seed(42)

    model = get_default_model()
    model.eval()

    inputs = get_sample_inputs()

    with torch.no_grad():
        output = model(*inputs)

    print('Input shape:', inputs[0].shape)
    print('Output shape:', output.shape if hasattr(output, 'shape') else 'N/A')
    print('Output value:', output)

    compiled_model = torch.compile(model)
    with torch.no_grad():
        output_compile = compiled_model(*inputs)

    print(f'Compile shape: {output_compile.shape if hasattr(output_compile, "shape") else "N/A"}')
    print(f'Compile value: {output_compile}')


if __name__ == '__main__':
    main()
```
output:
```
Input shape: torch.Size([8, 3, 28, 28])
Output shape: torch.Size([8, 10])
Output value: tensor([[ 0.1259,  2.1802, -0.4611,  0.2893, -0.2549,  1.3418,  1.6943, -0.3424,
         -1.1986, -0.6173],
        [-1.1378,  2.2973, -0.4874, -1.3212, -0.6088,  1.9096,  2.2077, -0.9895,
         -0.0292, -1.7046],
        [ 1.8430,  0.0883, -0.2677,  0.3203,  0.0061,  0.3310, -0.8915,  1.1340,
         -0.3330, -1.2369],
        [ 1.3185,  1.8414, -1.2385,  1.4909,  0.6500, -0.2228,  0.0992,  0.4742,
         -0.5263, -1.5285],
        [-0.1779, -0.3533,  0.2661, -0.0087, -2.2371, -1.3766,  0.1172,  0.4380,
          1.2886, -0.0678],
        [-0.1112, -1.1226,  0.4267,  0.1384,  0.1084,  0.0168,  0.4917,  0.9919,
          0.3643, -0.0441],
        [ 0.5743, -1.0891, -0.8450,  0.6866, -0.7328, -2.0010,  0.9617,  0.7150,
          1.1959,  2.1776],
        [-2.0769,  2.0931, -0.4595, -0.8991,  0.9147,  0.7922, -0.2049,  0.8360,
          2.1261, -1.2382]])
Compile shape: torch.Size([8, 10])
Compile value: tensor([[ 0.9969,  1.2417, -0.3827,  0.1177, -2.0274, -0.7088,  1.9214, -0.8019,
          0.1119,  1.5354],
        [-0.0927,  0.7070,  1.1071,  0.4636, -0.8059, -2.5171, -0.5633,  1.0416,
         -0.3674, -0.9421],
        [-0.8484,  0.6825, -0.7082, -0.4461, -2.0211,  0.7719, -0.0419,  1.2742,
          0.2612,  1.4381],
        [-0.1053,  1.1112,  1.4977,  0.6697, -1.0670, -0.1276, -0.3739, -0.2880,
          1.6734, -1.0643],
        [-0.3513,  0.5770,  1.5782, -1.1413, -0.9808, -1.2468,  1.3145,  1.5675,
          0.6031, -0.1875],
        [-0.4963, -0.7408,  0.6826,  0.8539,  2.2765,  0.4197,  1.8120,  1.1578,
         -1.5005,  1.9042],
        [-0.8276,  0.3552, -0.3252, -2.1116,  0.0803, -1.5034, -0.5395, -0.3793,
         -0.6101,  1.1910],
        [-1.2395, -0.3873,  0.4854, -1.0362, -0.5341,  0.1847,  0.5228, -0.7482,
         -0.6293,  0.5723]])
```

### Versions

Environment Information
PyTorch Build Details:

PyTorch version: 2.10.0.dev20251124+cpu

Is debug build: False

CUDA used to build PyTorch: Could not collect

ROCM used to build PyTorch: N/A

OS and Compilers:

OS: Ubuntu 24.04.1 LTS (x86_64)

GCC version: (Ubuntu 10.5.0-4ubuntu2) 10.5.0

Clang version: 18.1.3 (1)

CMake version: version 3.28.3

Libc version: glibc-2.39

Python Environment:

Python version: 3.12.3 (main, Nov 6 2025, 13:44:16) [GCC 13.3.0] (64-bit runtime)

Python platform: Linux-6.14.0-36-generic-x86_64-with-glibc2.39

Is CUDA available: False

CUDA runtime version: Could not collect

CUDA_MODULE_LOADING set to: N/A

GPU Information:

GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4060 Laptop GPU

Nvidia driver version: 580.95.05

cuDNN version: Could not collect

Is XPU available: False

HIP runtime version: N/A

MIOpen runtime version: N/A

Is XNNPACK available: True

Caching allocator config: N/A

CPU Information:

Architecture: x86_64

CPU op-mode(s): 32-bit, 64-bit

Address sizes: 39 bits physical, 48 bits virtual

Byte Order: Little Endian

CPU(s): 32

On-line CPU(s) list: 0-31

Vendor ID: GenuineIntel

Model name: Intel(R) Core(TM) i9-14900HX

CPU family: 6

Model: 183

Thread(s) per core: 2

Core(s) per socket: 24

Socket(s): 1

Stepping: 1

CPU(s) scaling MHz: 33%

CPU max MHz: 5800.0000

CPU min MHz: 800.0000

BogoMIPS: 4838.40

Flags: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect user_shstk avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi vnmi umip pku ospke waitpkg gfni vaes vpclmulqdq rdpid movdiri movdir64b fsrm md_clear serialize arch_lbr ibt flush_l1d arch_capabilities

Virtualization: VT-x

L1d cache: 896 KiB (24 instances)

L1i cache: 1.3 MiB (24 instances)

L2 cache: 32 MiB (12 instances)

L3 cache: 36 MiB (1 instance)

NUMA node(s): 1

NUMA node0 CPU(s): 0-31

Vulnerability Gather data sampling: Not affected

Vulnerability Ghostwrite: Not affected

Vulnerability Indirect target selection: Not affected

Vulnerability Itlb multihit: Not affected

Vulnerability L1tf: Not affected

Vulnerability Mds: Not affected

Vulnerability Meltdown: Not affected

Vulnerability Mmio stale data: Not affected

Vulnerability Reg file data sampling: Mitigation; Clear Register File

Vulnerability Retbleed: Not affected

Vulnerability Spec rstack overflow: Not affected

Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl

Vulnerability Spectre v1: Mitigation; usercopy/swapgs barriers and __user pointer sanitization

Vulnerability Spectre v2: Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S

Vulnerability Srbds: Not affected

Vulnerability Tsa: Not affected

Vulnerability Tsx async abort: Not affected

Vulnerability Vmscape: Mitigation; IBPB before exit to userspace

Versions of Relevant Libraries:

[pip3] numpy==2.3.3

[pip3] nvidia-cublas-cu12==12.1.3.1

[pip3] nvidia-cuda-cupti-cu12==12.1.105

[pip3] nvidia-cuda-nvrtc-cu12==12.1.105

[pip3] nvidia-cuda-runtime-cu12==12.1.105

[pip3] nvidia-cudnn-cu12==9.1.0.70

[pip3] nvidia-cufft-cu12==11.0.2.54

[pip3] nvidia-curand-cu12==10.3.2.106

[pip3] nvidia-cusolver-cu12==11.4.5.107

[pip3] nvidia-cusparse-cu12==12.1.0.106

[pip3] nvidia-nccl-cu12==2.21.5

[pip3] nvidia-nvjitlink-cu12==12.9.86

[pip3] nvidia-nvtx-cu12==12.1.105

[pip3] optree==0.18.0

[pip3] pytorch-triton-rocm==3.5.0

[pip3] torch==2.10.0.dev20251124+cpu

[pip3] torchao==0.15.0.dev20251124+cpu

[pip3] torchdata==0.12.0.dev20250909+cpu

[pip3] torchtext==0.17.0.dev20240912+cpu

[pip3] triton==3.1.0

[conda] Could not collect
